src/main/java/org/kohsuke/github/AbstractBuilder.java

```diff
@@ -0,0 +1,147 @@
+ * {@link #done()} when they are ready.
+ *
+ * @param <R>
+ *            Final return type for built by this builder returned when {@link #done()}} is called.
+ * @param <S>
+ *            Intermediate return type for this builder returned by calls to {@link #with(String, Object)}. If {@link S}
+ *            the same as {@link R}, this builder will commit changes after each call to {@link #with(String, Object)}.
+ */
+abstract class AbstractBuilder<R, S> {
+
+    @Nonnull
+    private final Class<R> returnType;
+
+    @Nonnull
+    private final Class<S> intermediateReturnType;
+
+    private final boolean commitChangesImmediately;
+
+    @CheckForNull
+    private final R baseInstance;
+
+    @Nonnull
+    protected final Requester requester;
+
+    // TODO: Not sure how update-in-place behavior should be controlled, but
+    // it certainly can be controlled dynamically down to the instance level or inherited for all children of some
+    // connection.
+    protected boolean updateInPlace;
+
+    /**
+     * Creates a builder.
+     *
+     * @param root
+     *            the GitHub instance to connect to.
+     * @param intermediateReturnType
+     *            the intermediate return type returned by calls to {@link #with(String, Object)}. Must either be equal
+     *            to {@code builtReturnType} or this instance must be castable to this class. If not, the constructor
+     *            {@link IllegalArgumentException}
+     * @param builtReturnType
+     *            the final return type for built by this builder returned when {@link #done()}} is called.
+     * @param baseInstance
+     *            optional instance on which to base this builder.
+     */
+    protected AbstractBuilder(@Nonnull GitHub root,
+            @Nonnull Class<S> intermediateReturnType,
+            @Nonnull Class<R> builtReturnType,
+            @CheckForNull R baseInstance) {
+        this.requester = root.createRequest();
+        this.returnType = builtReturnType;
+        this.intermediateReturnType = intermediateReturnType;
+        this.commitChangesImmediately = returnType.equals(intermediateReturnType);
+        if (!commitChangesImmediately && intermediateReturnType.isInstance(this)) {
+            throw new IllegalArgumentException(
+                    "Argument \"intermediateReturnType\": This instance must be castable to intermediateReturnType or intermediateReturnType must be equal to builtReturnType.");
+        }
+
+        this.baseInstance = baseInstance;
+        this.updateInPlace = false;
+
+    }
+
+    /**
+     * Finishes an update, committing changes.
+     *
+     * This method may update-in-place or not. Either way it returns the resulting instance.
+     *
+     * @return an instance with updated current data
+     * @throws IOException
+     *             if there is an I/O Exception
+     */
+    @Nonnull
+    public R done() throws IOException {
```

Comment:
Any suggestions for a better method name than `done()`?   `send()`?  `submit()`?  
We can't use `commit()` because it will almost surely collide with a git "commit" at least conceptually if not literally.

Reply:
BulkChange in core uses `commit` and `abort`. abort is not needed. Maybe `send` is a good choice. or `call`?

Reply:
The two places this will get used look something like this:
```java
GHLabel label = GHLabel.create().name(name).color(color).description(description).xxxx();
label.update().color(color2).description(description2).xxxx();
```
`send()` could work.  `call` not as much. I'd just like to make sure we've considered alternatives before releasing this.

Reply:
Okay, well, this api is internal, so changing it later is not a breaking API change.  I'll go with the current verbs for now and we can change them later if needed.

Reply:
Hm, doh, no, what was I thinking this is public.  Ah, but I can use the same structure we already have for API Previews and ship this now.  😄There we go.